modtools/fire-rate.lua

```diff
@@ -0,0 +1,511 @@
+    return weapon_data.fort_cap
+  else -- assume ADVENTURE
+    return weapon_data.adv_cap
+  end
+end
+
+function set_config_skill_mode(value)
+  if ( value ~= "basic" and value ~= "vanilla" ) then
+    config.skill_mode = "basic"
+    return
+  end
+
+  config.skill_mode = value
+end
+
+function get_config_skill_mode()
+  return config.skill_mode
+end
+
+---------------------------------------------------------------------
+-- Performs a skill roll based on configured mode
+function get_fire_rate_for_unit(unit, weapon_token, material_token)
+  local weapon_ranged_skill = get_weapon_skill_ranged(weapon_token)
+  local unit_skill_level = dfhack.units.getNominalSkill(
+                                                unit, weapon_ranged_skill, true)
+
+  local current_game_mode
+
+  -- Technically don't have to do this filtering, but it's good to do anyway
+  if ( dfhack.world.isFortressMode() ) then
+    current_game_mode = df.game_mode.DWARF
+  elseif ( dfhack.world.isAdventureMode() ) then
+    current_game_mode = df.game_mode.ADVENTURE
+  elseif ( dfhack.world.isArena() ) then
+    current_game_mode = df.global.gamemode
+  end
+
+  local fire_rate = get_weapon_fire_rate(weapon_token, material_token,
+                                         current_game_mode)
+  local speed_cap = get_weapon_fire_rate_cap(weapon_token, material_token,
+                                             current_game_mode)
+  local skill_factor = get_weapon_skill_factor(weapon_token, material_token,
+                                               current_game_mode)
+
+  -- Filter method based on current mode
+  -- (TODO: Except not actually because only one is available)
+  if ( get_config_skill_mode() == "basic" or true ) then
+    local uncapped_rate = math.floor(
+                            fire_rate - (unit_skill_level * skill_factor) + 0.5)
+
+    return math.max(uncapped_rate, speed_cap)
+  end
+end
+
+-- First rolls a skill check to determine how long it'll take for the unit to
+-- recover, then updates the unit's recovery time accordingly
+function do_shot_cooldown(unit, weapon_token, material_token)
+  local fire_rate = get_fire_rate_for_unit(unit, weapon_token, material_token)
+
+  apply_shot_cooldown(unit, fire_rate)
+end
+
+-- Causes the unit to have to wait the given length of time (in ticks) before
+-- they can act again
+-- This will account for the hard-coded delay present in adventure actions
+function apply_shot_cooldown(unit, recovery_time)
+  local time_to_use = recovery_time
+
+  -- Account for the hard-coded recovery time added for adventure mode actions
+  -- performed by the player (AI aren't subject to the delay)
+  if ( df.global.gamemode == df.game_mode.ADVENTURE
+            and unit.id == get_adventurer_unit().id ) then
+    -- The hardcoded delay is still active even in tactical mode, so no need to
+    -- try and account for it
+    time_to_use = math.max(recovery_time - 9, 0)
```

Comment:
It might be clearer to put the magic number 9 in a named constant so it's more obvious what is going on here

Reply:
That's a good idea. Made that edit.